### PR TITLE
[metal] use metal::fold_* instead of metal::accumulate

### PR DESCRIPTION
**Pull request by brunocodutra**

### CODE REVIEW

File: benchmark/type/fold_left/metal.list.cpp.erb [hunk 1]
```diff
@@ -13,12 +13,8 @@ template<int> struct x;
 
 <% env[:k].times do |k| %>
 
-using <%= "xs#{k}" %> = metal::list<
-    <%= ((k*n+1)..(k*n+n)).map { |i| "x<#{i}>" }.join(', ') %>
->;
-
 #if defined(METABENCH)
-using <%= "result#{k}" %> = metal::accumulate<metal::lambda<f>, state, <%= "xs#{k}" %>>;
+using <%= "result#{k}" %> = metal::fold_left<metal::lambda<f>, state<%= ((k*n+1)..(k*n+n)).map { |i| ", x<#{i}>" }.join('') %>>;
```
Comment (ldionne) [inline benchmark/type/fold_left/metal.list.cpp.erb]: Metal does not use a list anymore?

**Comment by brunocodutra** [comment]
metal::fold_left and metal::fold_right are defined on a variadic pack. For
lists, there is metal::accumulate, which left folds N lists in parallel
under a lambda that takes N + 1 arguments, but there is no equivalent of
right folding lists directly, hence the composition with metal::reverse
previously.

**Comment by ldionne** [comment]
OK, that addresses my concern.

File: benchmark/type/fold_left/metal.list.cpp.erb [hunk 2]
```diff
@@ -13,12 +13,12 @@ template<int> struct x;
 
 <% env[:k].times do |k| %>
 
-using <%= "xs#{k}" %> = metal::list<
-    <%= ((k*n+1)..(k*n+n)).map { |i| "x<#{i}>" }.join(', ') %>
+using <%= "xs#{k}" %> = meta::list<
```
Comment (ldionne) [inline benchmark/type/fold_left/metal.list.cpp.erb]: I think this should be `metal::list`, not `meta::list`.

Comment (brunocodutra) [inline benchmark/type/fold_left/metal.list.cpp.erb]: Oops

**Comment by brunocodutra** [comment]
this time properly tested

**Comment by ldionne** [comment]
How did one of the Travis job fail but the other one succeed?

Also, nitpick, but when you submit PRs from branches that you push to this GitHub repository, please prefix them with `pr/`. Otherwise, it creates two Travis jobs (one for the pull request, and one for the branch push itself).

**Comment by brunocodutra** [comment]
> How did one of the Travis job fail but the other one succeed?

Didn't both pass?

> when you submit PRs from branches that you push to this GitHub repository, please prefix them with pr/

Sure, I keep forgetting about that, sorry.

**Comment by ldionne** [comment]
> Didn't both pass?

No, one of them failed (but then I restarted it and it seems to have passed, oh well). I thought C++ compilation was deterministic.



> Sure, I keep forgetting about that, sorry.

No worries.

**Pull request merged**